enhancements/20240805-encrypted-block-volume.md

```diff
@@ -0,0 +1,440 @@
+# Support Block Volume Encryption/Decryption
+
+This enhancement focuses on the encryption and decryption of block volumes, particularly involving boot volumes and data volumes. 
+
+
+## Summary
+
+This enhancement introduces encryption and decryption capabilities for block volumes, specifically targeting boot volumes (virtual machine images) and data volumes. This functionality will enhance security by ensuring that data stored in these volumes is encrypted and can be decrypted when attached. 
+
+## Related Issues
+
+https://github.com/harvester/harvester/issues/3129
+
+## Motivation
+
+
+### Goals
+
+- Implement encryption for boot volumes and data volumes (support hot plug). 
+- Provide a clear and detailed setup process for users to enable encryption/decryption. 
+
+### Non-goals
+
+- This enhancement will not cover GUI-related changes.
+- **This enhancement only aims to admin user since Harvester only has an admin user and lacks of user management.**
+
+
+## Proposal
+
+Since Longhorn uses clone mechanism to encrypt/decrypt Boot volume, so Harvester will follow that. This means that there would be two virtual machine image after encryption. One is encrypted, and the other is source virtual machine image. 
+
+![](./20240805-encrypted-block-volume/lh-flow.png)
+
+Data volumes are different from Boot volume. Because Longhorn supports block encryption/decryption, so Harvester could encrypt/decrypt data volume directly. 
+
+Regardless of volumes used, Longhorn will decrypt volume when it's attached to virtual machine. So, we'll be able to read the boot volumes and data volumes in the virtual machine. 
+
+### User Stories
+
+#### Story 1 - Use encrypted boot volume to start virtual machine
+
+Admin wants to encrypt the current boot volume for security, so Admin follows up the guideline:
+
+1. Prepare a virtual machine image, which you could upload or use URL to download.
+2. Create a secret for encryption and template storage class.
+3. Create a new virtual machine image with encryption enabled.
+
+After that, Admin gets an encrypted virtual machine image as boot volume. Then, Admin can start the virtual machine with selecting encrypted boot volume in GUI. 
+
+For security reason, Admin decides to delete that original virtual machine image. The encrypted image is independent of the original image. 
+
+For this moment, Admin only has this encrypted virtual machine image and Admin can reuse this boot volume.
+
+Over time, Admin wants to decrypt the encrypted virtual machine image. Admin follows up the guideline:
+
+1. Prepare an encrypted virtual machine image.
+2. Create a new virtual machine image with decryption enabled.
+
+After that, Admin will have an encrypted virtual machine image and an unencrypted virtual machine image. Admin could select one of them to start the virtual machine in GUI.
+
+
+#### Story 2 - Use encrypted data volume in virtual machine
+
+Admin wants to encrypt a data volume for security, so Admin follows up the guideline:
```

Comment:
Admin prepares the secret and storage class. But any user who creates volumes from the storage class can enjoy the encryption at rest feature. The paragraph can be clarified that a non-admin user still can use the encryption.

Reply:
I added one sentence in the end 

> Furthermore, any non-admin users in Rancher Integration mode can still use this storage class to create an encrypted volume.